### ISSUE TITLE
[Bug]: Compose-related memory leaks

### ISSUE THREAD

**Issue by brian-livefront** [android/nowinandroid#1446]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is there a StackOverflow question about this issue?

- [X] I have searched StackOverflow

### What happened?

The `SearchViewModel` (and likely other ViewModels) appears to leak after navigating to the search screen and back to the main screen. This seems to be related to [this Compose issue](https://issuetracker.google.com/issues/312895362).

**Steps to reproduce**
- Open the app.
- Navigate to the search screen.
- Enter text in the search field.
- Exit the search screen.
- Open the memory profiler, attempt to force a garbage collection several times, and then collect a heap dump.
- Search for `SearchViewModel` in the heap.

The `SearchViewModel` will still be present in the heap even though the screen was left and many garbage collections were attempted.

<img width="800" alt="Screenshot 2024-05-16 at 12 50 46 PM" src="https://github.com/android/nowinandroid/assets/126027223/0b2e2c67-502b-4cd3-8e20-639fc1f2a297">

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by sl-ahmedabdellaoui** [comment]
Hello,

I have encountered the same issue with TopicViewModel. Here are the steps to reproduce it:

- Navigate to Interests.
- Open one Topic.
- Go back to Interests.
- Return to the Home screen.
- Open a profiler, record a dump, and search for TopicViewModel.

You'll notice that TopicViewModel is still present. The memory leak appears to be related to a lambda function referencing the ViewModel. Upon further inspection, it seems the ViewModel is retained by the AccessibilityManager as shown below:

![Capture d’écran 2024-07-26 à 16 08 12](https://github.com/user-attachments/assets/b1954769-0803-4df9-87c7-36da873edff4)

Is there any update on this issue ?

By the way, even in another Android app using Compose, I noticed the same behavior. ViewModels seem to be kept in memory even after `onCleared` is called.
thank you

**Comment by carterinchina** [comment]
yes  I have encountered the same issue  too

**Comment by Mihail-Rain-Heart** [comment]
yes, I see the same with `TopicViewModel` on `main` branch with last update `13.11.2025`

upd: I see https://github.com/android/nowinandroid/issues/581#issuecomment-1425675689

I do in `NiaAppState.kt`
```kotlin
    fun navigateToTopLevelDestination(topLevelDestination: TopLevelDestination) {
        trace("Navigation: ${topLevelDestination.name}") {
            val topLevelNavOptions = navOptions {
                // Pop up to the start destination of the graph to
                // avoid building up a large stack of destinations
                // on the back stack as users select items
                popUpTo(navController.graph.findStartDestination().id) {
                    saveState = false
                }
                // Avoid multiple copies of the same destination when
                // reselecting the same item
                launchSingleTop = true
                // Restore state when reselecting a previously selected item
                restoreState = false
            }

            when (topLevelDestination) {
                FOR_YOU -> navController.navigateToForYou(topLevelNavOptions)
                BOOKMARKS -> navController.navigateToBookmarks(topLevelNavOptions)
                INTERESTS -> navController.navigateToInterests(null, topLevelNavOptions)
            }
        }
    }
```
Yes I don't keep data and remind previous state by reset `saveState` and `restoreState` flags.
I can't understand. Navigation it`s always pain. If this api leaked, so why demo app use it? My heart is broken.
